experimental/unicodeset_parser/src/parse.rs

```diff
@@ -0,0 +1,1192 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::fmt::Display;
+use std::{collections::HashSet, iter::Peekable, str::CharIndices};
+
+use icu_collections::{
+    codepointinvlist::{CodePointInversionList, CodePointInversionListBuilder},
+    codepointinvliststringlist::CodePointInversionListAndStringList,
+};
+use icu_properties::maps::{load_general_category, load_script};
+use icu_properties::provider::*;
+use icu_properties::sets::{load_for_ecma262_unstable, CodePointSetData};
+use icu_properties::{GeneralCategory, Script};
+use icu_provider::prelude::*;
+
+/// The kind of error that occurred.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ParseErrorKind {
+    /// An unexpected character was encountered. This variant implies the other variants
+    /// (notably `UnknownProperty` and `Unimplemented`) do not apply.
+    UnexpectedChar(char),
+    /// The property name or value is unknown. For property names, make sure you use the spelling
+    /// defined in [ECMA-262](https://tc39.es/ecma262/#table-nonbinary-unicode-properties).
+    UnknownProperty,
+    /// The source is an incomplete unicode set.
+    Eof,
+    /// Something unexpected went wrong with our code. Please file a bug report on GitHub.
+    Internal,
+    /// The provided syntax is not supported by us. Note that unknown properties will return the
+    /// `UnknownProperty` variant, not this one.
+    Unimplemented,
+    /// The provided escape sequence is not a valid Unicode code point.
+    InvalidEscape,
+}
+use ParseErrorKind as PEK;
+
+impl ParseErrorKind {
+    fn with_offset(self, offset: usize) -> ParseError {
+        ParseError {
+            offset: Some(offset),
+            kind: self,
+        }
+    }
+}
+
+impl From<ParseErrorKind> for ParseError {
+    fn from(kind: ParseErrorKind) -> Self {
+        ParseError { offset: None, kind }
+    }
+}
+
+/// The error type returned by the `parse` functions in this crate.
+///
+/// See [`ParseError::fmt_with_source`] for pretty-printing and [`ParseErrorKind`] of the
+/// different types of errors represented by this struct.
+#[derive(Debug, Clone)]
+pub struct ParseError {
+    // offset is the index to an arbitrary byte in the last character in the source that makes sense
+    // to display as location for the error, e.g., the unexpected character itself or
+    // for an unknown property name the last character of the name.
+    offset: Option<usize>,
+    kind: ParseErrorKind,
+}
+
+type Result<T, E = ParseError> = core::result::Result<T, E>;
+
+impl ParseError {
+    /// Pretty-prints this error and if applicable, shows where the error occurred in the source.
+    ///
+    /// Must be called with the same source that was used to parse the set.
```

Comment:
Thought: This seems like it's probably a reasonable way to avoid problems associated with holding a reference to the string in the error struct. On the other hand, you could return `Result<CPILSL, ParseError<'_>>` (borrowing the string) and if people want to un-borrow the string they can call some function that drops the reference.